src/mtype.c

```diff
@@ -3980,6 +3980,8 @@ void TypeSArray::resolve(Loc loc, Scope *sc, Expression **pe, Type **pt, Dsymbol
     }
     else
     {
+        if ((*pt)->ty != Terror)
+            next = *pt;     // prevent re-running semantic() on 'next'
```

Comment:
I'm not sure what's the hackish.

If a part of a type (eg. TypeSArray::next) is determined to be valid (== semantic analysis is succeeded),
to store back it again in the part of whole expression (eg. TypeSArray) is completely valid.

In `Type(SArray|Darray|AArray|Pointer|Delegate)::semantic()`,  same logic is actually used.

```cpp
Type *TypeSArray::semantic(Loc loc, Scope *sc)
{
    ...
    Type *tn = next->semantic(loc, sc);
    if (tn->ty == Terror)
        return terror;

    ...
    this->next = tn;
    ...
    t = this->addMod(tn->mod);
    return t->merge();
}
```

On the other hand, if `next->deco` is alredy set, `next->resolve(.., &t, ..)` must return the `next` itself to `t`.
It's guaranteed by the `Type::merge()`logic, so it's definitely reliable.